rust/operator-binary/src/controller/apply.rs

```diff
@@ -0,0 +1,146 @@
+use crate::OPERATOR_NAME;
+
+use snafu::{OptionExt, ResultExt, Snafu};
+use stackable_druid_crd::{DruidCluster, APP_NAME};
+use stackable_operator::{
+    client::Client,
+    cluster_resources::ClusterResources,
+    k8s_openapi::api::core::v1::Secret,
+    kube::{runtime::controller::Action, Resource, ResourceExt},
+    role_utils::RoleGroupRef,
+};
+use std::sync::Arc;
+use strum::{EnumDiscriminants, IntoStaticStr};
+
+use super::{types::AppliableClusterResource, CONTROLLER_NAME};
+
+#[derive(Snafu, Debug, EnumDiscriminants)]
+#[strum_discriminants(derive(IntoStaticStr))]
+#[allow(clippy::enum_variant_names)]
+pub enum Error {
+    #[snafu(display("failed to apply global Service"))]
+    ApplyRoleService {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to create cluster resources"))]
+    CreateClusterResources {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to apply Service for {}", rolegroup))]
+    ApplyRoleGroupService {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<DruidCluster>,
+    },
+    #[snafu(display("failed to apply discovery ConfigMap"))]
+    ApplyDiscoveryConfig {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to apply ConfigMap for {}", rolegroup))]
+    ApplyRoleGroupConfig {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<DruidCluster>,
+    },
+    #[snafu(display("failed to apply StatefulSet for {}", rolegroup))]
+    ApplyRoleGroupStatefulSet {
+        source: stackable_operator::error::Error,
+        rolegroup: RoleGroupRef<DruidCluster>,
+    },
+    #[snafu(display("object defines no namespace"))]
+    ObjectHasNoNamespace,
+    #[snafu(display("failed to retrieve secret for internal communications"))]
+    FailedToRetrieveInternalSecret {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to delete orphaned resources"))]
+    DeleteOrphanedResources {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to apply internal secret"))]
+    ApplyInternalSecret {
+        source: stackable_operator::error::Error,
+    },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub async fn handle_cluster_resources(
+    client: &Client,
+    druid: &Arc<DruidCluster>,
+    appliable_cluster_resources: Vec<AppliableClusterResource>,
```

Comment:
I think you can do away with `AppliableClusterResource` and just use `ClusterResource`:
```suggestion
    appliable_cluster_resources: Vec<ClusterResource>,
```
Then this function just loops over the array and applies each element. And then it's just a generic function in operator-rs. It's even future-proof if we add more `ClusterResource` types.

Reply:
There are upsides and downsides to both choices. I went with enums to easily pass additional information along with the `ClusterResource` and keep scope of necessary code changes at repo level.